pkg/porter/copy.go

```diff
@@ -4,63 +4,67 @@ import (
 	"fmt"
 	"strings"
 
-	"github.com/docker/distribution/reference"
+	"get.porter.sh/porter/pkg/cnab"
 	"github.com/pkg/errors"
 )
 
 type CopyOpts struct {
 	Source           string
+	sourceRef        cnab.OCIReference
 	Destination      string
 	InsecureRegistry bool
 }
 
 // Validate performs validation logic on the options specified for a bundle copy
 func (c *CopyOpts) Validate() error {
+	var err error
 	if c.Destination == "" {
 		return errors.New("--destination is required")
 	}
-	source, err := reference.ParseNormalizedNamed(c.Source)
+
+	c.sourceRef, err = cnab.ParseOCIReference(c.Source)
 	if err != nil {
 		return errors.Wrap(err, "invalid value for --source, specified value should be of the form REGISTRY/bundle:tag or REGISTRY/bundle@sha")
 	}
-	if isCopyDigestReference(source) && isCopyReferenceOnly(c.Destination) {
+	if c.sourceRef.HasDigest() && isCopyReferenceOnly(c.Destination) {
 		return errors.New("--destination must be tagged reference when --source is digested reference")
 	}
 	return nil
 }
 
-func isCopyDigestReference(source reference.Named) bool {
-	if _, ok := source.(reference.Canonical); ok {
-		return true
-	}
-	return false
-}
-
 func isCopyReferenceOnly(dest string) bool {
-	named, err := reference.ParseNormalizedNamed(dest)
+	ref, err := cnab.ParseOCIReference(dest)
 	if err != nil {
 		return false
 	}
-	return reference.IsNameOnly(named)
+	return ref.IsRepositoryOnly()
 }
 
-func generateNewBundleRef(source, dest string) string {
+func generateNewBundleRef(source cnab.OCIReference, dest string) cnab.OCIReference {
 	if isCopyReferenceOnly(dest) {
-		bundleNameRef := source[strings.LastIndex(source, "/")+1:]
-		return fmt.Sprintf("%s/%s", dest, bundleNameRef)
+		srcVal := source.String()
+		bundleNameRef := srcVal[strings.LastIndex(srcVal, "/")+1:]
+		dest = fmt.Sprintf("%s/%s", dest, bundleNameRef)
 	}
-	return dest
+	destRef, err := cnab.ParseOCIReference(dest)
+	if err != nil {
+		panic(err)
```

Comment:
Want to double-check that we want to keep the panic here, only b/c we usually avoid doing so (besides unit test helpers, etc.)

Reply:
This function is supposed to panic. You're right that I don't use them elsewhere in the code. This function is only for tests, so that you can parse a well known reference in a test and it will hard fail if you mess it up, without having to remember to check the err. It also simplifies the return arguments so that you can do things like this

```
assert.Equal(t, MustParseOCIReference("getporter/porter-hello:v0.1.1"), myref)
```

if we used the other function that would look like this

```
gotref, err := ParseOCIReference("getporter/porter-hello:v0.1.1")
require.NoErr(t, err)
assert.Equal(t, gotref, myref)
```

The function name starts with Must, which is a pattern for functions like this that will panic when there is an error.

Reply:
Wait, but I'm looking at the following logic in `generateNewBundleRef` which is called during `CopyBundle`, i.e. non-test code, right?

```
	destRef, err := cnab.ParseOCIReference(dest)
	if err != nil {
		panic(err)
	}
```

Just want to triple-check 😉 .  I did see `MustParseOCIReference` being used in tests -- good idea.

Reply:
MustParseOCIReference (the one that can panic) is only used in tests. The other, cnab.ParseOCIReference, is used in real code but we always check the error returned. Sorry let me know if I'm missing something, totally not trying to blow you off here!

Reply:
Gotcha.  Porter will panic on corner cases of `porter copy`; just want to make sure we're okay with that.  e.g.

```
$ p copy --source localhost:5000/nfs:v0.1.0 --destination oops/
panic: invalid reference format oops/: invalid reference format

goroutine 1 [running]:
get.porter.sh/porter/pkg/porter.generateNewBundleRef(0x2e3a168, 0xc000341e40, 0x7ffeefbff39d, 0x5, 0x105344f, 0x12)
	/Users/vdice/go/src/get.porter.sh/porter/pkg/porter/copy.go:51 +0x246
get.porter.sh/porter/pkg/porter.(*Porter).CopyBundle(0xc0004d7110, 0xc0001999c0, 0x0, 0x0)
	/Users/vdice/go/src/get.porter.sh/porter/pkg/porter/copy.go:58 +0x65
main.buildBundleCopyCommand.func2(0xc00039a780, 0xc000199b80, 0x0, 0x4, 0x0, 0x0)
	/Users/vdice/go/src/get.porter.sh/porter/cmd/porter/copy.go:29 +0x33
github.com/spf13/cobra.(*Command).execute(0xc00039a780, 0xc000199b40, 0x4, 0x4, 0xc00039a780, 0xc000199b40)
	/Users/vdice/go/pkg/mod/github.com/spf13/cobra@v1.1.3/command.go:852 +0x472
github.com/spf13/cobra.(*Command).ExecuteC(0xc000473900, 0xc000473900, 0x101a801, 0x0)
	/Users/vdice/go/pkg/mod/github.com/spf13/cobra@v1.1.3/command.go:960 +0x375
github.com/spf13/cobra.(*Command).Execute(...)
	/Users/vdice/go/pkg/mod/github.com/spf13/cobra@v1.1.3/command.go:897
main.main()
	/Users/vdice/go/src/get.porter.sh/porter/cmd/porter/main.go:23 +0x65
```

Reply:
Oh! Thanks I see what you are pointing out now. Let me fix that